src/internal.c

```diff
@@ -776,19 +794,231 @@ union wolfSSH_key {
+ * @return          keyId as int, WS_MEMORY_E, WS_UNIMPLEMENTED_E
+ */
+static int IdentifyKey(const byte* in, word32 inSz, int isPrivate, void* heap)
+{
+    union wolfSSH_key *key = NULL;
+    int keyId = ID_UNKNOWN;
+    word32 scratch;
+    int ret = WS_SUCCESS;
+    int dynType = isPrivate ? DYNTYPE_PRIVKEY : DYNTYPE_PUBKEY;
+
+    key = (union wolfSSH_key*)WMALLOC(sizeof(union wolfSSH_key), heap, dynType);
+    if (key == NULL) {
+        ret = WS_MEMORY_E;
+    }
+
+#ifndef WOLFSSH_NO_RSA
+    if (ret == WS_SUCCESS) {
+        /* Check RSA key */
+        if (keyId == ID_UNKNOWN) {
+            scratch = 0;
+            ret = wc_InitRsaKey(&key->rsa, NULL);
+
+            if (ret == 0) {
+                if (isPrivate) {
+                    ret = wc_RsaPrivateKeyDecode(in, &scratch,
+                            &key->rsa, inSz);
+                }
+                else {
+                    ret = wc_RsaPublicKeyDecode(in, &scratch,
+                            &key->rsa, inSz);
+                }
+
+                /* If decode was successful, this was an RSA key. */
+                if (ret == 0) {
+                    keyId = ID_SSH_RSA;
+                }
+                /* in case we need to check for another key flavor */
+                ret = WS_SUCCESS;
+            }
+
+            wc_FreeRsaKey(&key->rsa);
+        }
+    }
+#endif
+#ifndef WOLFSSH_NO_ECDSA
+    if (ret == WS_SUCCESS) {
+        if (keyId == ID_UNKNOWN) {
+            scratch = 0;
+            ret = wc_ecc_init_ex(&key->ecc, heap, INVALID_DEVID);
+
+            if (ret == 0) {
+                if (isPrivate) {
+                    ret = wc_EccPrivateKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+                else {
+                    ret = wc_EccPublicKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+
+                /* If decode was successful, this was an RSA key. */
+                if (ret == 0) {
+                    switch (wc_ecc_get_curve_id(key->ecc.idx)) {
+                        case ECC_SECP256R1:
+                            keyId = ID_ECDSA_SHA2_NISTP256;
+                            break;
+                        case ECC_SECP384R1:
+                            keyId = ID_ECDSA_SHA2_NISTP384;
+                            break;
+                        case ECC_SECP521R1:
+                            keyId = ID_ECDSA_SHA2_NISTP521;
+                            break;
+                    }
+                }
+                /* in case we need to check for another key flavor */
+                ret = WS_SUCCESS;
+            }
+
+            wc_ecc_free(&key->ecc);
+        }
+    }
+#endif /* ! WOLFSSH_NO_ECDSA */
+
+    if (key != NULL) {
+        WFREE(key, heap, dynType);
+    }
+
+    if (keyId == ID_UNKNOWN) {
+        ret = WS_UNIMPLEMENTED_E;
+    }
+
+    return ret == WS_SUCCESS ? keyId : ret;
```

Comment:
I think it would be more readable if we add an else to the if at like 894 and set ret to keyId. Then just return ret without the ternary operator.

Reply:
Check